avi_model_tests.ipynb

```diff
@@ -0,0 +1,13890 @@
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "Clearly the images aren't the same size, so I am applying transforms to resize and also normalize.\n"
+      ],
+      "metadata": {
+        "id": "S9ufqufcxWBb"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "\n",
+        "transform = transforms.Compose([\n",
+        "    transforms.Resize((256, 256)),  # Match the expected input size\n",
+        "    transforms.ToTensor(),\n",
+        "    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])  # Normalization\n",
+        "])\n"
+      ],
+      "metadata": {
+        "id": "PdZ6wxmn2ZJf"
+      },
+      "execution_count": 16,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "### Creating the dataset and dataloader"
+      ],
+      "metadata": {
+        "id": "6InheyNnxesB"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "train_dataset = CaptionDataset(train_df, tokenizer, transform)\n",
+        "val_dataset = CaptionDataset(val_df, tokenizer, transform)\n",
+        "test_dataset = CaptionDataset(test_df, tokenizer, transform)\n",
+        "\n",
+        "train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True, collate_fn=collate_fn)\n",
+        "val_loader = DataLoader(val_dataset, batch_size=32, collate_fn=collate_fn)\n",
+        "test_loader = DataLoader(test_dataset, batch_size=32, shuffle=False, collate_fn=collate_fn)"
+      ],
+      "metadata": {
+        "id": "y3UfyWVm2ahO"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Model 1: CNN and LSTM"
+      ],
+      "metadata": {
+        "id": "iLJulh-0KBeX"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "### Model Class"
+      ],
+      "metadata": {
+        "id": "APZ2yi3L-z98"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "I am using a CNN to extract features from images (ResNet50) and then using an LSTM to generate captions"
+      ],
+      "metadata": {
+        "id": "qnEjEB_VKD8T"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "class ImageCaptioningModel(nn.Module):\n",
+        "    def __init__(self, embed_dim, hidden_dim, vocab_size, num_layers):\n",
+        "        super(ImageCaptioningModel, self).__init__()\n",
+        "\n",
+        "        # load pretrained resnet model\n",
+        "        resnet = resnet50(pretrained=True)\n",
+        "\n",
+        "        # freeze the resnet layers so that they are not trained\n",
+        "        for param in resnet.parameters():\n",
+        "            param.requires_grad = False\n",
+        "\n",
+        "        # remove the classification layer and replace it with a linear layer\n",
```

Comment:
Comments are very in depth and helfup to explain your model structure here, thanks!